source/common/upstream/subset_lb.cc

```diff
@@ -101,6 +112,30 @@ SubsetLoadBalancer::~SubsetLoadBalancer() {
   });
 }
 
+void SubsetLoadBalancer::rebuildSingle() {
+  if (single_) {
+    // Because PriorityUpdateCb doesn't give a modified list (only added and removed), it is
+    // faster to just rebuild this map than try to figure out if any hosts had their metadata
+    // changed, and then figure out the old and new value (to remove from the old key in this map
+    // and insert in the new key).
+    single_host_per_subset_map_.clear();
+    for (const auto& host_set : original_priority_set_.hostSetsPerPriority()) {
+      for (const auto& host : host_set->hosts()) {
+        MetadataConstSharedPtr metadata = host->metadata();
+        const auto& filter_metadata = metadata->filter_metadata();
+        auto filter_it = filter_metadata.find(Config::MetadataFilters::get().ENVOY_LB);
+        if (filter_it != filter_metadata.end()) {
+          const auto& fields = filter_it->second.fields();
+          auto fields_it = fields.find(single_key_);
+          if (fields_it != fields.end()) {
```

Comment:
Can we surface duplicates to the control plane somehow?

Reply:
I was going to at least implement a metric. We could also consider throwing, which I think would reject the EDS update entirely, but I'd prefer just a metric I think.